src/main.rs

```diff
@@ -26,10 +33,76 @@ fn main() {
     }
 }
 
+fn run_in_userns() -> anyhow::Result<()> {
+    // dnf needs to be run as root, but given that rpmoci only needs to query package repos
+    // and/or install packages into an install root, we can run in a user namespace, mapping
+    // the current uid/gid to root
+    // this function spawns a child process in a new user namespace, the parent configures
+    // the uid/gid mappings, then signals the child to re-exec rpmoci
+
+    let user_id = getuid();
+    let group_id = getgid();
+    // dnf chooses a root-writable cache directory by default. It's unlikely that
+    // the current user will be able to write there, so we configure the RPMOCI_CACHE_DIR
+    // to point to a cache directory in the current user's home directory.
+    let cache_dir = dirs::cache_dir().unwrap().join("rpmoci");
+
+    const STACK_SIZE: usize = 1024 * 1024;
+    let stack: &mut [u8; STACK_SIZE] = &mut [0; STACK_SIZE];
+    // this pipe is for the parent to notify the child when user namespaces mappings
+    // have been configured
+    let (reader, writer) = pipe()?;
+    // create child process with a new user namespace
+    let child = nix::sched::clone(
+        Box::new(|| {
+            // this child process just waits for the parent to notify it before re-execing
+            close(writer).unwrap();
+            read(reader, &mut Vec::new()).unwrap();
+            Command::new(std::env::current_exe().unwrap())
+                .args(std::env::args().skip(1))
+                .env("RPMOCI_CACHE_DIR", cache_dir.clone())
+                .exec();
+            255
+        }),
+        stack,
+        CloneFlags::CLONE_NEWUSER,
```

Comment:
Throw in a mount-namespace too?  I'm wondering what things RPM installers might try to do, and I could believe that mounting a new FS might be on that list.

Theoretically you could add a PID namespace too, but hopefully not (that has to be done differently, by `unshare`-ing in the parent before the `clone` (PID namespaces can't be changed in a process, only the namespace for spawned children is changed, so the new process picks it up).  If you do add a PID NS you'll _also_ want to re-mount `/proc` in the child (you can just mount it on top of the existing one, since your in a mount NS) so that PIDs and `/proc/[pid]` line up.

Reply:
I added mount namespace too on balance - even if it's not needed it's unlikely that user will be able to create user namespace but not mount